lib/tree.rb

```diff
@@ -16,43 +20,45 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
+  def add(key, value = nil)
+    @root = add_under_root(key, value, @root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(log(n))
   def find(key)
```

Comment:
👍 The time/space complexities are O(log n) if the tree is balanced and O(n) if it's not balanced.